### ISSUE TITLE
Hidden string length parameter missing in Fortran interface?

### ISSUE THREAD

**Issue by mkre** [eschnett/MPIwrapper#16]
Hi,

First of all, I am not really a Fortran person, so please excuse my ignorance. I am trying to adopt the MPItrampoline+MPIwrapper toolchain but ran into an issue with a Fortran library (MUMPS). It crashed with the following backtrace:
```
#0  0x00007ffff49af019 in __strncpy_avx2 () from target:/lib64/libc.so.6
#1  0x00007fffeb9f9477 in ompi_fortran_string_c2f (
    cstr=0x7fffffff30b0 "di2denbg0056lpc.net.plm.eds.com",
    fstr=0x7fffffff32a0 "di2denbg0056lpc.net.plm.eds.com", len=-52580) at strings.c:94
#2  0x00007fffeb6ecbd2 in ompi_get_processor_name_f (
    name=0x7fffffff32a0 "di2denbg0056lpc.net.plm.eds.com", resultlen=0x7fffffff329c,
    ierr=0x7fffffff330c, name_len=-52580) at pget_processor_name_f.c:94
#3  0x00007fffebd22da5 in mpiabi_get_processor_name_ (
    name=0x7fffffff32a0 "di2denbg0056lpc.net.plm.eds.com", resultlen=0x7fffffff329c,
    ierror=0x7fffffff330c)
    at /u/ydfb4q/tpl/mpiwrapper/build-openmpi-4/src/mpiabi_defn_functions_fortran.h:4684
#4  0x00007fff6b218aae in mumps_build_arch_node_comm (comm=9, newcomm=0, newsize=0, newrank=0)
    at src/tools_common.F:1572
```
The issue is clearly that `name_len=-52580`, so the string copy fails. After a bit of investigation, the issue seems to be the "hidden string length" parameters when calling C functions from Fortran. Open MPI somehow [adds this parameter explicitly](https://github.com/open-mpi/ompi/blob/v4.1.5/ompi/mpi/fortran/mpif-h/get_processor_name_f.c#L79) to their `mpif.h` API, but MPIwrapper is missing the parameter, leading to the bogus negative value and finally the crash. 

Changing the [corresponding block in `mpi_functions_fortran.py`](https://github.com/eschnett/MPIwrapper/blob/v2.10.1/mpiabi/mpi_functions_fortran.py#L1855) to
```
    ("void", "MPI_Get_processor_name", [
        ("char *", "name"),
        ("int *", "resultlen"),
        ("MPI_Fint *", "ierror"),
        ("size_t", "name_len"),
    ]),
```
avoids the crash. However, there are more places in the code where similar measure might need to be taken. The type of the additional parameter (`int` or `size_t`) [depends on the compiler and its version](https://gcc.gnu.org/onlinedocs/gcc-9.2.0/gfortran/Argument-passing-conventions.html).

Do you think we also need these compiler-dependent parameters in MPIwrapper? If so, would we also need them in MPItrampoline? My current workaround only requires to change MPIwrapper, but not MPItrampoline, but I guess we would probably need to change both to be consistent.

The similar project WI4MPI also [seems to inject these parameters into the API](https://github.com/cea-hpc/wi4mpi/blob/master/generator/generate.py#L547), but again, I am not really a Fortran expert so I might be missing something obvious.

Thanks,
Moritz

**Comment by eschnett** [comment]
This is resolved in the new versions of MPIwrapper 2.10.2 / MPItrampoline 5.2.2.

**Comment by eschnett** [comment]
No, I think there is an error in my implementation (although it passes tests).

**Comment by eschnett** [comment]
Thank you for your report. This is resolved in the new versions of MPIwrapper 2.10.3 / MPItrampoline 5.2.3.

**Comment by mkre** [comment]
Thanks for addressing this so quickly!
